### DISCUSSION TITLE
Will the Waifu2x browser plugin be created?

### DISCUSSION THREAD

**Discussion by My12123** [nagadomi/waifu2x#446] [Ideas]
Will an extension be created for Google Chrome? Available at https://chrome.google.com/webstore/category/extensions

**Comment by nagadomi** [comment]
I do not plan to create a chrome-extension at this time. However, there are several ways to implement it.

#### 1. Jump to waifu2x.udp.jp with URL Query string
waifu2x.udp.jp supports specifying parameters with query string. So it can be called by specifying the URL of the image from the right-click menu.
Example
https://waifu2x.udp.jp/?url=https://upload.wikimedia.org/wikipedia/commons/6/6d/Simple_Wikipe_tan_by_Kasuga39.png&style=art&scale=2&noise=0

Pros: Easy to implement
Cons: Direct conversion is not allowed unless the user is a Patreon supporter (due to reCaptcha). Also, for some reason, the image URL may not be accessible.

#### 2. In-browser processing (ONNX Runtime Web)
Using [unlimited.waifu2x](https://unlimited.waifu2x.net/) code, users can process images directly (inplace replace) on a web page.

Cons: Currently, only CPU mode is supported, so it is very slow

**Reply by My12123** [reply]
If you use the local version http://127.0.0.1:8812/index.ru.html will it work?

**Reply by nagadomi** [reply]
If web.py supports Cross-Origin Resource Sharing, API is available through XHR from chrome-extension.
However, I don't think anyone use it, so I won't develop it.

**Reply by My12123** [reply]
I think for those who read manga and comics, this will do.

**Reply by nagadomi** [reply]
WebGPU will be enabled in google-chrome 113, scheduled for release this month.
Then you can run ONNX models on GPU in the browser without local server setup.

**Reply by My12123** [reply]
Can I get more details?

**Reply by nagadomi** [reply]
WebGPU is a new API that allows GPU to be used from JavaScript in the browser.
Currently, google-chrome does not have this feature enabled by default. But it will be enabled in version 113, which will be released this month.
See https://developer.chrome.com/blog/webgpu-release/

Also, ONNX Runtime Web, a library for running ONNX models in web browser, is currently developing WebGPU backend.
unlimited.waifu2x.net uses WebAssembly backend(CPU) of ONNX Runtime Web, so once WebGPU backend is available, it will be able to run on GPU.

These are still uncertain, but the run-time environment for AI in the browser could change dramatically in the coming months.
